File: Python/Lesson 1/BinaryGap.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(N):
    # Implement your solution here
    stack = []
    while N != 0:
        if N % 2 == 0:
            stack.append(0)
        else:
            stack.append(1)
        N //= 2
    startCounting = False
    maxGap = 0
    currentGap = 0
    while len(stack) != 0:
        if startCounting:
            if stack.pop() == 0:
                currentGap += 1
            else:
                if currentGap > maxGap:
                    maxGap = currentGap
                currentGap = 0
        if not startCounting and stack.pop() == 1:
            startCounting = True

    return maxGap

File: C++/Lesson 1/BinaryGap.cpp
// you can use includes, for example:
// #include <algorithm>
#include <stack>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(int N) {
    // Implement your solution here
    stack<int> stack;
    while (N != 0) {
        if (N % 2 == 0) {
            stack.push(0);
        } else {
            stack.push(1);
        }
        N /= 2;
    }
    int maxBinaryGap = 0;
    int currentBinaryGap = 0;
    bool countGap = false;
    while (!stack.empty()) {
        if (stack.top() == 1)
        {
            if (countGap = true) {
                if (currentBinaryGap > maxBinaryGap) {
                    maxBinaryGap = currentBinaryGap;
                }
                currentBinaryGap = 0;
            } else {
                countGap = true;
            }
        } else {
            if (countGap) {
                currentBinaryGap++;
            } 
        }
        stack.pop();
    }  
    return maxBinaryGap;
}